### PR TITLE
Documentation for using MSBuildTaskBase class

### PR DESCRIPTION
**Pull request by missymessa**
Documentation to give instruction on best practices for using the MSBuildTaskBase class and how it works behind the scenes.

### CODE REVIEW

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Purpose of MSBuildTaskBase Class
+
+In an effort to provide dependency injection for the MSBuild Tasks in Arcade, we wanted to have a consistent way in implementing dependency injection for the `Microsoft.Build.Utilities.Task` type. In a web service or command line application, the entry point into the program serves as a place to configure the dependency injection service collections and provider. Since there isn't an entry point like that for MSBuild Tasks, we have created an abstract class to handle the common elements of configuriation.
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:3]: I'd probably rephrase this a little to explain the "effort" is to enabled testing and ease code reuse.  DI is just the mechanism we are using to reach that goal.  So that you don't have "but why DI at all" questions hindering adoption. :-)  If that's better explained in some other doc, a link to it and a short blurb would be good.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+1. In your MSBuild Task class, inherit the `MSBuildTaskBase` class found the in `Microsoft.Arcade.Common` project. 
+
+```
+public class TaskClassName : MSBuildTaskBase
+{
+    ...
+}
+```
+
+2. Impelment the following two methods: `ExecuteTask` and `ConfigureServices`.
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:18]: Step 2 sounds like "do step 3 and 4".  Does step 2 need to exist, or can we just combine this info into steps 3 and 4.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    ...
+}
+```
+
+2. Impelment the following two methods: `ExecuteTask` and `ConfigureServices`. 
+
+    - `ExecuteTask` replaces the `Execute` method from Microsoft.Build.Utilities.Task object. Any functionality you would include in the `Execute` method should go into `ExecuteTask` instead. 
+    - `ConfigureServices` should contain the configuration for the Service Collection that will resolve the dependencies
+
+3. Similarly to how interfaces are included into a constructor for a class that will have it's dependencies injected into it, include the interfaces of the dependencies for the Task into `ExecuteTask` parameters.
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:23]: I would be specific with the requirements here it:
* Must be public
* Must be named exactly "ExecuteTask"
* Return a boolean
* Can accept any (or no) paramters, which will be resolved from the service collection configured by ConfigureServices

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+```
+
+4. Similarly to how a service collection would be configured for how to resolve dependencies, `ConfigureServices` should contain the same kind of configuration. Ensure that you have configured all dependencies, 
+
+```
+public override void ConfigureServices(IServiceCollection collection)
+{
+    collection.TryAddSingleton<IThing, ThingClass>();
+    collection.TryAddSingleton<IFoo, FooClass>();
+    ...
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:39]: Is there a "base" call necessary here?  Or does the base class handle that a different way?  If so, we should include it in the example in the correct place.

**Review by MattGal — APPROVED** [review]
Good stuff.  Does seem to assume the reader attended your brown bag.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+In an effort to provide dependency injection for the MSBuild Tasks in Arcade, we wanted to have a consistent way in implementing dependency injection for the `Microsoft.Build.Utilities.Task` type. In a web service or command line application, the entry point into the program serves as a place to configure the dependency injection service collections and provider. Since there isn't an entry point like that for MSBuild Tasks, we have created an abstract class to handle the common elements of configuriation. 
+
+Link to the [MSBuildTaskBase](https://github.com/dotnet/arcade/blob/master/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs) class.
+
+# How to Implement in a MSBuild Task
+
+1. In your MSBuild Task class, inherit the `MSBuildTaskBase` class found the in `Microsoft.Arcade.Common` project. 
+
+```
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:11]:
```suggestion
```csharp
```
It will add syntax highlighting

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+{
+    MSBuildTaskClass task = new MSBuildTaskClass();
+
+    var collection = new ServiceCollection();
+    task.ConfigureServices(collection);
+    var provider = collection.BuildServiceProvider();
+
+    foreach(var dependency in task.GetExecuteParameterTypes())
+    {
+        var service = provider.GetRequiredService(dependency);
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:61]: This will cause them to be constructed.

There should be an optional parameter to "IsDependencyResolutionCoherent" that is "additionalTypes".  We can just pass those types to it.  (It makes for nicer error messages as well)

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+```
+MSBuildTaskClass task = new MSBuildTaskClass();
+task.ConfigureServices(collection);
+```
+
+3. Build the service provider and pass it into the method that will resolve the dependencies and invoke the `ExecuteTask` method.
+
+```
+using var provider = collection.BuildServiceProvider();
+task.InvokeExecute(provider);
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:100]: Probably want to save the output boolean. :-)

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+# What's Happening Behind the Scenes
+
+The `MSBuildTaskBase` class inherits `Microsoft.Build.Utilities.Task`, so it implements the `Execute` method that MSBuild Task classes implement. It uses reflection to accomplish resolving the services defined in the concrete class' `ExecuteTask` method. 
+
+The Task's `Execute` is traditionally the method that is called when running the task. It is required to be implemented when inheriting the `Microsoft.Build.Utilities.Task` class. 
+
+The `Execute` method serves as the Task's entry point, that handles the registration of dependencies that are defined in the concrete implementation of the `MSBuildTaskBase` class. In this method, a new `ServiceCollection` is created. It is passed into the `ConfigureServices` method to be populated with the configuration defined in the concrete implementation. Then, it builds the service provider that is passed into the `InvokeExecute` method. 
+
+```
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:111]: Could this just be a link to the actual code?  Code in docs go stale _so fast_ that if it's real code in the codebase, it's better to just reference it.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+2. Impelment the following two methods: `ExecuteTask` and `ConfigureServices`. 
+
+    - `ExecuteTask` replaces the `Execute` method from Microsoft.Build.Utilities.Task object. Any functionality you would include in the `Execute` method should go into `ExecuteTask` instead. 
+    - `ConfigureServices` should contain the configuration for the Service Collection that will resolve the dependencies
+
+3. Similarly to how interfaces are included into a constructor for a class that will have it's dependencies injected into it, include the interfaces of the dependencies for the Task into `ExecuteTask` parameters. 
+
+```
+public bool ExecuteTask(IThing thing, IFoo foo, IBar bar)
```
Comment (ChadNedzlek) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: Hrm... I wonder if this needs to be (or could optionally be) async.  A lot of things build tasks need to do will hit networks/files, and many of the modern SDK's don't have non-async versions of thsoe API's, so it's easiest if the entry point can just be async. (And even if it isn't the cost of a non-async async method is very low, since if it never yields and returns synchronously, then it dosen't have to invoke the async state machine)


We could be generous and look for _either_ `bool ExecuteTask` OR `Task<bool> ExecuteTaskAsync`.  I'm curious if we find that lots of people that implement this need to wrap the async stuff themselves, and we could alleviate that a bit.

**Review by ChadNedzlek — APPROVED** [review]

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
+# Purpose of MSBuildTaskBase Class
+
+In an effort to provide dependency injection for the MSBuild Tasks in Arcade, we wanted to have a consistent way in implementing dependency injection for the `Microsoft.Build.Utilities.Task` type. In a web service or command line application, the entry point into the program serves as a place to configure the dependency injection service collections and provider. Since there isn't an entry point like that for MSBuild Tasks, we have created an abstract class to handle the common elements of configuriation. 
+
+Link to the [MSBuildTaskBase](https://github.com/dotnet/arcade/blob/master/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs) class.
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:5]: Nit: It's weird that the link is on the name of the file and not on the word "link." So suggesting a rewording here.
```suggestion
MSBuildTaskBase class can be found [here](https://github.com/dotnet/arcade/blob/master/src/Microsoft.Arcade.Common/MSBuildTaskBase.cs).
```

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+# How to Implement in a MSBuild Task
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:7]: nit: should be ##

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+3. Similarly to how interfaces are included into a constructor for a class that will have it's dependencies injected into it, include the interfaces of the dependencies for the Task into `ExecuteTask` parameters. 
+
+```
+public bool ExecuteTask(IThing thing, IFoo foo, IBar bar)
+{ 
+    ...
+}
+```
+
+4. Similarly to how a service collection would be configured for how to resolve dependencies, `ConfigureServices` should contain the same kind of configuration. Ensure that you have configured all dependencies,
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:32]: Nit:

```suggestion
4. Similarly to how a service collection would be configured to resolve dependencies, `ConfigureServices` should contain the same kind of configuration. Ensure that you have configured all dependencies, 
```

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+public class TaskClassName : MSBuildTaskBase
+{
+    ...
+}
+```
+
+2. Impelment the following two methods: `ExecuteTask` and `ConfigureServices`. 
+
+    - `ExecuteTask` replaces the `Execute` method from Microsoft.Build.Utilities.Task object. Any functionality you would include in the `Execute` method should go into `ExecuteTask` instead. 
+    - `ConfigureServices` should contain the configuration for the Service Collection that will resolve the dependencies
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:21]: Nit: missing a period at the end of this sentence

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 3 above)
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:23]: This sentence is really hard to parse. I'm trying to come up with a rewrite that gets the point across, but I'm struggling. Maybe something like:

```suggestion
3. `ExecuteTask` is similar to a constructor that has its dependencies injected. Similar to how that constructor would take its injected dependencies as parameters, `ExecuteTask` should take its injected dependencies as parameters.
```

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+```
+public override void ConfigureServices(IServiceCollection collection)
+{
+    collection.TryAddSingleton<IThing, ThingClass>();
+    collection.TryAddSingleton<IFoo, FooClass>();
+    ...
+}
+```
+
+# How to Use with Tests
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:43]: Nit: ##

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+## Validating Dependency Injection Configuration
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:45]: Nit: ###

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            out string message
+        )
+        .Should()
+        .BeTrue(message);
+}
+```
+
+This test uses a helper class that is provided in the `Microsoft.DotNet.Internal.DependencyInjection.Testing` project. 
+
+## Mocking Services for Task Unit Tests
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:79]: Nit: ###

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
+# Purpose of MSBuildTaskBase Class
```
Comment (adiaaida) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:1]: So I've left this comment a bunch of times, so I'm just going to summarize my thoughts on the #'s in md files. Single # is for the Title of the doc, ## is for Section headers, ### is for subsection headers, etc. Basically, everything needs an extra #.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 4 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:39]: In my implementations, I haven't used the base call, just overrode the virtual method so I had more control over the configuration. I can definitely mention it here in case folks want to use the base in their implementations.

Comment (alexperovich) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:39]: does the base actually do anything? It probably shouldn't.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (alexperovich) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: I would prefer this be async.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+1. Create the mocks for the services to be injected into the task, and add them to a new service collection to be used in the test. 
+
+```csharp
+var collection = new ServiceCollection()
+    .AddSingleton(thingMock.Object)
+    .AddSingleton(fooMock.Object);
+```
+
+2. Pass that collection into the Task's `ConfigureServices` method. If the `ConfigureServices` method registers dependencies using TryAdds, it shouldn't add conflicting dependency registrations into the collection.
```
Comment (alexperovich) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]: I think this is backwards. The Task's ConfigureServices method needs to add everything it needs to execute, so by definition it will add non-mocks for everything. The mocks need to be added after calling ConfigureServices so they replace the default implementations.

Although I think a test that is testing the task's functionality shouldn't even call ConfigureServices, just invoke the ExecuteTask method and pass in mocks to the parameters.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 4 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:39]: Right now, it configures the IFileSystem and Log object, since those seem to be common dependencies. I expect that as we continue to refactor, we may find more common dependencies to add to that, or maybe not? Since this is pretty new, we're learning as we're going on what we'll need or not. I don't mind striking the method in the base, either and turning it into an abstract method instead of a virtual one.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 8 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Agreed. Let me see if there's a fancy markdown way I can embed the actual code instead of having to copy/paste it.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+}
+```
+
+`MSBuildTaskBase` contains a virtual implementation of `ConfigureServices` that configures common dependencies, such as `IFileSystem` (if you want to abstract the File and Directory classes so they can be mocked out for tests) and `TaskLoggingHelper` object called `Log`. It is expected that this method will be overridden. 
+
+```csharp
+public virtual void ConfigureServices(IServiceCollection collection)
+{
+    collection.TryAddSingleton<IFileSystem, FileSystem>();
+    collection.TryAddSingleton(Log);
```
Comment (alexperovich) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:129]: I'm not a fan of the "default" services getting added in the base implementation of ConfigureServices. In asp.net the ConfigureServices method doesn't have to call base, because the base method doesn't do anything, the "default" services are configured before calling ConfigureServices. I can't see any task not wanting these services, so every task needs to call base.ConfigureServices(). That is different from how asp.net works so it will probably cause confusion.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 8 above)
Comment (riarenas) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: If the code lives in the same repo as the PR, just using a link to the lines you care about should do it. Sort of like here: 

https://github.com/dotnet/arcade/blob/cac51b21e00714e9a51b63f4a4a51bf92a4685b1/eng/Signing.props#L25-L28

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 18 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]: You're right, a test doesn't even need to call ConfigureServices if it's mocking out everything. I can definitely mention that as an option, and that if the developer doesn't care about certain mocks and wants the default service, they can pass their collection into the task's ConfigureServices method to get the stuff they didn't want/need to configure.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 19 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:129]: @ChadNedzlek, any opposition for me changing ConfigureServices to an abstract method?

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 8 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: perfect, that's exactly the magic I want

Comment (riarenas) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Ooh, but I'm not sure if it works in checked in docs. I hope it does!

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: With all the refactoring I've done so far, I haven't seen any Executes being called with the async keyword. Might be worth investigating for the future if there's a requirement. 

@mmitche and @chcosta, thoughts?

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 4 above)
Comment (alexperovich) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:39]: I think an abstract method is better, with default dependencies added before calling it would be easier. That makes the ConfigureServices behave the same as the one in ServiceHost, and asp.net core.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (mmitche) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: We do have instances of ExecuteAsync tasks:

https://github.com/dotnet/arcade-services/blob/bf6ee9dbe20c94cdc95938673f82936613ff8692/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs#L62-L71

The root thing msbuild calls is never async

Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: Not that it probably matters, but that example is in Arcade Services, instead of Arcade. But if MSBuild itself never calls an async Execute, then we probably won't ever need it?

**Review by garath — APPROVED** [review]

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (chcosta) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: Supporting both doesn't cost much.  It's up to the author of the task to implement any asynchronous behavior though, and I agree with this

> even if it isn't the cost of a non-async async method is very low, since if it never yields and returns synchronously, then it dosen't have to invoke the async state machine

Async opens up more performant coding patterns, even if, at the end of the day, the root MSBuild call is synchronous.

I'm ambivalent with only a slight preference towards async.  I can't think of anything I want to write (at the moment), that needs async, but I also don't see the harm and generally prefer the pattern.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 18 above)
Comment (ChadNedzlek) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]: I think they should get in the habit of calling "ConfigureServices" even if they do mock everything. I've added wrappers before that don't really need to get mocked (to isolate and share some piece of business logic, for example).

It's _super_ frustrating to have to go edit dozens or hundreds of tests if they aren't calling the base registration methods.

It's a cheap thing to do in the tests that doesn't hurt anything, and can prevent a lot of pain later.

I'd leave the document calling it.

As to the before/after, it only really matters if someone is resolving an IEnumerable<IThing>.  If it's before, then the "try" won't register another one, and the mock will "win"... but that also probably defeats the purpose of the IEnumerable.

Basically you can't mock things that are injected by the IEnumerable<> interface.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 19 above)
Comment (ChadNedzlek) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:129]: No, makes perfect sense.  Probably wrap it in some _other_ method... so in the base class you have something like:
```csharp
public bool TaskExecuterThing() {
    // make the collection
    InitalizeServiceCollection(collection);
    using var provider = collection.BuildServiceProvider();
}

public void InitalizeServiceCollection(IServiceCollection collection) {
    ConfigureServices(collection);
    collection.TryAddSingleton<IFileSystem, FileSystem>();
    collection.TryAddSingleton(Log);
}

protected abstract void ConfigureServices(IServiceCollection collection);
```

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 18 above)
Comment (alexperovich) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]: technically you can, but you would have to remove things from the collection, which isn't super hard, just annoying

Comment (ChadNedzlek) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]: True, true. Hadn't considered that you can actually remove stuff.  It's... weird.  But it could be done if necessary. :-)

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+3. Similarly to how interfaces are included into a constructor for a class that will have its dependencies injected into it, include the interfaces of the dependencies for the Task into `ExecuteTask` parameters. 
+
+```csharp
+public bool ExecuteTask(IThing thing, IFoo foo, IBar bar)
+{ 
+    ...
+}
+```
+
+4. Similarly to how a service collection would be configured to resolve dependencies, `ConfigureServices` should contain the same kind of configuration. Ensure that you have configured all dependencies,
```
Comment (lukas-lansky) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:32]:
```suggestion
4. Similarly to how a service collection would be configured to resolve dependencies, `ConfigureServices` should contain the same kind of configuration. Ensure that you have configured all dependencies.
```

**Review by lukas-lansky — APPROVED** [review]

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+1. In your MSBuild Task class, inherit the `MSBuildTaskBase` class found the in `Microsoft.Arcade.Common` project. 
+
+```csharp
+public class TaskClassName : MSBuildTaskBase
+{
+    ...
+}
+```
+
+2. Implement the following two methods: `ExecuteTask` and `ConfigureServices`.
```
Comment (riarenas) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:18]: [nit] I'm a fan of sticking to using 1. for every item in a markdown numbered list and letting the parser show the right numbers. If I need to add a step in between two existing items, I don't have to change the numbers in every step afterwards.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (riarenas) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:26]: I too have a slight preference towards making async part of the guidelines. I remember a few instances in buildtools where we were bitten by not having an async implementation when we needed it. The pattern has usually been for MSBuild to call the synchronous execute, and the synchronous execute calls into the async method.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 18 above)
Comment (riarenas) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:89]:
```suggestion
2. Pass that collection into the Task's `ConfigureServices` method. If the `ConfigureServices` method registers dependencies using `TryAdd`s, it shouldn't add conflicting dependency registrations into the collection. 
```

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+    ServiceCollection collection = new ServiceCollection();
+    ConfigureServices(collection);
+    using (var provider = collection.BuildServiceProvider())
+    {
+        return InvokeExecute(provider);
+    }
+}
+```
+
+`MSBuildTaskBase` contains a virtual implementation of `ConfigureServices` that configures common dependencies, such as `IFileSystem` (if you want to abstract the File and Directory classes so they can be mocked out for tests) and `TaskLoggingHelper` object called `Log`. It is expected that this method will be overridden.
```
Comment (riarenas) [inline Documentation/Mechanics/MSBuildTaskDependencyInjection.md:123]: [nit]
```suggestion
`MSBuildTaskBase` contains a virtual implementation of `ConfigureServices` that configures common dependencies, such as `IFileSystem` (if you want to abstract the File and Directory classes so they can be mocked out for tests) and a `TaskLoggingHelper` object called `Log`. It is expected that this method will be overridden. 
```

**Review by riarenas — APPROVED** [review]
I like this! mostly nits, and perhaps some sections could benefit from some more context to be more helpful for first time readers.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 21 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:18]: This isn't working, I think, because of the breaks where the code snippets are :(

Comment (riarenas) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:18]: I think it's due to indentation. Your code blocks should probably be intended to show that they are part of the list item

1. This
```c#
var x = 0
```
1. doesn't work

but 

1. this
    ```c#
    var x = 0
    ```
1. does

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 1 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Reworded this to open with more information about making MSBuild Tasks more unit-testable.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 2 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Combined this info into steps 3 and 4 (now steps 2 and 3).

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 3 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Added these bullet points to this step.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Finished what I meant to say :)

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 3 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Reworded this a bit.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+public class TaskClassName : MSBuildTaskBase
+{
+    ...
+}
+```
+
+2. Implement the following two methods: `ExecuteTask` and `ConfigureServices`. 
+
+    - `ExecuteTask` replaces the `Execute` method from Microsoft.Build.Utilities.Task object. Any functionality you would include in the `Execute` method should go into `ExecuteTask` instead. 
+    - `ConfigureServices` should contain the configuration for the Service Collection that will resolve the dependencies.
```
[the remark this answers is not in the document]
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:29]: Grabbed the links that Chad included in his DI presentation add included them on this page.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 24]
```diff
@@ -0,0 +1,140 @@
+# Purpose of MSBuildTaskBase Class
+
+In an effort to provide dependency injection for the MSBuild Tasks in Arcade, we wanted to have a consistent way in implementing dependency injection for the `Microsoft.Build.Utilities.Task` type. In a web service or command line application, the entry point into the program serves as a place to configure the dependency injection service collections and provider. Since there isn't an entry point like that for MSBuild Tasks, we have created an abstract class to handle the common elements of configuration.
```
[the remark this answers is not in the document]
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:5]: Clarified this by talking about how we want to make MSBuild Tasks more unit-testable

**Comment by missymessa** [comment]
> Good stuff. Does seem to assume the reader attended your brown bag.

Added the links from Chad's DI presentation for further reading about dependency injection.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 9 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Created https://github.com/dotnet/arcade/issues/6772 to address the inclusion of turning this into an async method.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 19 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md:138]: Included this refactor in this issue: https://github.com/dotnet/arcade/issues/6772

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 18 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: kk, resolving this comment. Leaving everything as is :)

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 6 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Refactored this in all the tests and the documentation. Definitely handier to use the additionalScopedTypes than this.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 8 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Embedding doesn't work in docs like it does here :(  Included links to the code in the document where relevant.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+                task.ConfigureServices(s);
+            },
+            out string message
+        )
+        .Should()
+        .BeTrue(message);
+}
+```
+
+This test uses a helper class that is provided in the `Microsoft.DotNet.Internal.DependencyInjection.Testing` project.
```
[the remark this answers is not in the document]
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Included this remark in the whole bit about writing this test.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 4 above)
Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Turned this method into an abstract method. Added a bit in the enhancement issue to re-evaluate necessary default services that should be configured.

**Comment by missymessa** [comment]
Whew. Okay, I think I got through everyone's feedback (thank you all!) and created another issue to add some enhancements to the base class (e.g. asyncing the ExecuteTask method; including default services where necessary; improving the logger so that we can use ILogger types). Also did a little bit of refactoring to the tests that I created to make better use of the DI testing functionality. 

I'll merge this PR sometime tomorrow afternoon if there's no other feedback.

File: Documentation/Mechanics/MSBuildTaskDependencyInjection.md (hunk 6 above)
Comment (ChadNedzlek) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: Hrm.  I don't _think_ we actually scope them, so it's probably more accurate to use the "additionalSingletonTypes".  The two parameters are mostly just "is this type going to be resolved inside a nested scope or not".

The only difference is that "additionalSingletonTypes" will report an error if you resolve a "AddScoped" object, since it's implied they are being used in a non-scoped context, and that does weird/bad things in DI.

Comment (missymessa) [reply Documentation/Mechanics/MSBuildTaskDependencyInjection.md]: I updated it to Singleton after I made this comment. I realized my mistake and then verified that you had a Scoped AND Singleton version of the additional types in there :)
